rfcs/000-third-party-software-installation.md

```diff
@@ -0,0 +1,109 @@
+- Topic Name: `third-party-software-installation`
+- Start Date: 2022-03-22
+- RFC PR: [vlang/rfcs#0000](https://github.com/vlang/rfcs/pull/22)
+  <!-- - V Issue: [vlang/v#00000](https://github.com/vlang/v/issues/00000) -->
+
+# Summary
+
+Allow installation of third-party V software.
+This goes beyond just placing the source code in `VMODULES`.
+It allows building the software from source locally and adding it to path for the user to uses from anywhere.
+
+Additionally, we propose the use of `~/.v` and the relevant subfolders (like `~/.v/modules`, `~/.v/bin`) to store the source code, built binaries, etc.
+
+# Motivation
+
+Currently, V only allows downloading the source code for V modules in the `VMODULES` directory through the `v install` command.
+This is good for making modules globally available. There is currently no streamlined way of _installing_ or _fetching_ or _obtaining_ third-party software written in V.
+
+# Guide-level explanation
+
+Take [_klonol_](https://github.com/hungrybluedev/klonol) for example.
+It is a command-line tool written in V that helps to maintain a local copy of all git repositories on GitHub and Gitea.
+If you want to install it, you have to clone the source yourself, run `v build.vsh` and then add the `bin` folder to `PATH`.
+A nice thing to have is a `v fetch <repo>` or `v obtain <repo>` command (or something else, please comment) that will clone the repo and build it for you.
+
+Only once, like when V is install, for example, the user needs to add `~/.v/bin` to `PATH`.
+This will make all third-party applications installed using `v fetch <repo>` available from anywhere.
+
+# Reference-level explanation
+
+It will piggyback off of `v install`.
+The act of installing the software is done through the following additional steps:
+
+1. Running `v build.vsh` or some other specified build script in the root of the repo.
+2. This will create an executable, along with all other
+   relevant files (like templates, etc) in the `bin` directory of the repo. Ideally, the resources will be embedded in the executable to allow for only one executable to be enough.
+3. The executable will be added to `~/.v/bin` for the user to call from anywhere.
+   TODO.
+   An additional **suggestion** is to have the `~/.vmodules` directory be replaced in favour of `~/.v/modules`.
+
+If there is another script (other than `build.vsh`) that generates the executable
+
+A sample `build.vsh` script is:
```

Comment:
inside the script we should include the bug report, for instance, if there is some error during the build, we should store the cmd result, and ask the user if want to report it on Github.

In addition, we can do static analysis on the compiler error and may suggest some hotfix.

Reply:
We should make a list of proper workflow that 3rd party projects need to follow. Like for example:
1. Obtaining source code
2. Checking formatting.
3. Removing old artifacts.
4. Build executable.
5. Report error if it fails by making a log file.

etc.

Reply:
I agree!

However, some of them can be included in some standard rules, that we can discover only by implementing them.